File: apps/sandbox-ts/process/broadcaster.ts
export class Broadcaster<T>{
    private subscribers = new Set<(data:T) => void>()
    private closed = false;
    private onCloseHandlers = new Set<() => void>()

    subscribe(fn: (data: T) => void): () => void {
        if(this.closed){
            throw new Error("Broadcaster closed");
        }

        this.subscribers.add(fn)

        return () => {
            this.subscribers.delete(fn)
        }
    }

    publish(data: T){
        if(this.closed) return;

        for(const fn of this.subscribers){
            try {
                fn(data)
            } catch (error) {
                
            }
        }
    }

    onClose(fn: () => void):() => void{
        if(this.closed){
            fn()
            return () => {}
        }
        this.onCloseHandlers.add(fn)
        return () => this.onCloseHandlers.delete(fn)
    }

    close(){
        if(this.closed) return;
        this.closed = true;
        
        for(const fn of this.onCloseHandlers) fn();
        this.subscribers.clear();
        this.onCloseHandlers.clear()
    }
}

File: apps/sandbox-go/internal/process/broadcaster.go
package process

import "sync"

type Broadcaster struct{
	mu sync.Mutex
	subscribers map[chan []byte]struct{}
	closed bool
}

func NewBroadcaster() *Broadcaster{
	return &Broadcaster{
		subscribers: make(map[chan []byte]struct{}),
	}
}

func (b *Broadcaster) Subscribe() (<-chan []byte, func()){
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.closed {
		ch := make(chan []byte)
		close(ch)
		return ch, func() {}
	}

	ch := make(chan []byte, 16)
	b.subscribers[ch] = struct{}{}

	unsub := func ()  {
		b.mu.Lock()
		delete(b.subscribers, ch)
		close(ch)
		b.mu.Unlock()
	}

	return ch, unsub
}

func(b *Broadcaster) Publish(data []byte){
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.closed{
		return
	}

	for ch := range b.subscribers {
		select{
		case ch <- data:
		default:
		}
	}
}

func (b *Broadcaster) Close(){
	b.mu.Lock()
	defer b.mu.Unlock()

	if b.closed{
		return
	}
	b.closed = true

	for ch := range b.subscribers{
		close(ch)
	}
	b.subscribers = nil
}